src/highway.zig

```diff
@@ -0,0 +1,311 @@
+
+/// Result structure for character finding operations
+pub const IndexResult = extern struct {
+    index: i32, // -1 if not found
+    count: i32,
+};
+
+extern "c" fn highway_char_frequency(
+    text: [*]const u8,
+    text_len: usize,
+    freqs: [*]i32,
+    delta: i32,
+) void;
+
+extern "c" fn highway_index_of_char(
+    haystack: [*]const u8,
+    haystack_len: usize,
+    needle: u8,
+) usize;
+
+extern "c" fn highway_index_of_interesting_character_in_string_literal(
+    noalias text: [*]const u8,
+    text_len: usize,
+    quote: u8,
+) usize;
+
+extern "c" fn highway_index_of_newline_or_non_ascii(
+    noalias haystack: [*]const u8,
+    haystack_len: usize,
+) usize;
+
+extern "c" fn highway_index_of_newline_or_non_ascii_or_ansi(
+    noalias haystack: [*]const u8,
+    haystack_len: usize,
+) usize;
+
+extern "c" fn highway_index_of_newline_or_non_ascii_or_hash_or_at(
+    noalias haystack: [*]const u8,
+    haystack_len: usize,
+) usize;
+
+extern "c" fn highway_index_of_space_or_newline_or_non_ascii(
+    noalias haystack: [*]const u8,
+    haystack_len: usize,
+) usize;
+
+extern "c" fn highway_contains_newline_or_non_ascii_or_quote(
+    noalias text: [*]const u8,
+    text_len: usize,
+) bool;
+
+extern "c" fn highway_index_of_needs_escape_for_javascript_string(
+    noalias text: [*]const u8,
+    text_len: usize,
+    quote_char: u8,
+) usize;
+
+extern "c" fn highway_index_of_any_char(
+    noalias text: [*]const u8,
+    text_len: usize,
+    noalias chars: [*]const u8,
+    chars_len: usize,
+) usize;
+
+extern "c" fn highway_fill_with_skip_mask(
+    mask: [*]const u8,
+    mask_len: usize,
+    output: [*]u8,
+    input: [*]const u8,
+    length: usize,
+    skip_mask: bool,
+) void;
+
+/// Count frequencies of [a-zA-Z0-9_$] characters in a string
+/// Updates the provided frequency array with counts (adds delta for each occurrence)
+pub fn scanCharFrequency(text: string, freqs: *[64]i32, delta: i32) void {
+    if (text.len == 0 or delta == 0) {
+        return;
+    }
+
+    highway_char_frequency(
+        text.ptr,
+        text.len,
+        freqs.ptr,
+        delta,
+    );
+}
+
+pub fn indexOfChar(haystack: string, needle: u8) ?usize {
+    if (haystack.len == 0) {
+        return null;
+    }
+
+    const result = highway_index_of_char(
+        haystack.ptr,
+        haystack.len,
+        needle,
+    );
+
+    if (result == haystack.len) {
+        return null;
+    }
+
+    bun.debugAssert(haystack[result] == needle);
+
+    return result;
+}
+
+pub fn indexOfInterestingCharacterInStringLiteral(slice: string, quote_type: u8) ?usize {
+    if (slice.len == 0) {
+        return null;
+    }
+
+    const result = highway_index_of_interesting_character_in_string_literal(
+        slice.ptr,
+        slice.len,
+        quote_type,
+    );
+
+    if (result == slice.len) {
+        return null;
+    }
+
+    return result;
+}
+
+pub fn indexOfNewlineOrNonASCII(haystack: string) ?usize {
```

Comment:
Most of these functions need their names updated. They will return non-null for more than they say

Reply:
i'll rename it to be indexOfInterestingCharacter